storage/libradb/src/pruner/mod.rs

```diff
@@ -0,0 +1,238 @@
+    num_versions_to_keep: u64,
+    worker_thread: Option<JoinHandle<()>>,
+    command_sender: Mutex<Sender<Command>>,
+}
+
+impl Pruner {
+    /// creates a worker thread that waits on a channel for pruning commands.
+    pub fn new(db: Arc<DB>, num_versions_to_keep: u64) -> Self {
+        let (command_sender, command_receiver) = channel();
+
+        let worker_thread =
+            std::thread::spawn(move || Worker::new(db, command_receiver).work_loop());
+
+        Self {
+            num_versions_to_keep,
+            worker_thread: Some(worker_thread),
+            command_sender: Mutex::new(command_sender),
+        }
+    }
+
+    /// Sends pruning command to the worker thread when necessary.
+    pub fn wake(&self, latest_version: Version) {
+        if latest_version > self.num_versions_to_keep {
+            let least_readable_version = latest_version - self.num_versions_to_keep;
+            match self.command_sender.lock() {
+                Ok(sender) => sender
+                    .send(Command::Prune {
+                        least_readable_version,
+                    })
+                    .expect("Receiver should not destruct prematurely."),
+                Err(e) => crit!("Failed locking command_sender to pruner worker. {:?}", e),
+            }
+        }
+    }
+
+    /// (For tests only.) Notifies the worker thread and waits for it to finish its job by loop
+    /// querying an internal counter.
+    #[cfg(test)]
+    pub fn wake_and_wait(&self, latest_version: Version) -> Result<()> {
+        // To prevent interference between tests in the same process, set it to 0 beforehand.
+        OP_COUNTER.set(COUNTER_NAME, 0);
+        self.wake(latest_version);
+
+        if latest_version > self.num_versions_to_keep {
+            let least_readable_version = latest_version - self.num_versions_to_keep;
+            const TIMEOUT: Duration = Duration::from_millis(100);
+            let end = Instant::now() + TIMEOUT;
+
+            while Instant::now() < end {
+                if OP_COUNTER.gauge(COUNTER_NAME).get() >= least_readable_version as i64 {
+                    return Ok(());
+                }
+                sleep(Duration::from_millis(1));
+            }
+            bail!("Timeout waiting for pruner worker.");
+        }
+        Ok(())
+    }
+}
+
+impl Drop for Pruner {
+    fn drop(&mut self) {
+        self.command_sender
+            .lock()
+            .expect("Locking command_sender should not fail.")
+            .send(Command::Quit)
+            .expect("Receiver should not destruct.");
+        self.worker_thread
+            .take()
+            .expect("Worker thread must exist.")
+            .join()
+            .expect("Worker thread should join peacefully.");
+    }
+}
+
+enum Command {
+    Quit,
+    Prune { least_readable_version: Version },
+}
+
+struct Worker {
+    db: Arc<DB>,
+    command_receiver: Receiver<Command>,
+    least_readable_version: Version,
+    blocking_recv: bool,
```

Comment:
Might be worth adding a bit comment about `blocking_recv`, so people can understand before reading `work_loop`.